src/types/complex-types/array.ts

```diff
@@ -250,85 +250,116 @@ function reconcileArrayChildren<T>(
-                    fail(
-                        `Cannot add an object to a state tree if it is already part of the same or another state tree. Tried to assign an object to '${parent.path}/${subPath}', but it lives already at '${childNode.path}'`
-                    )
-                }
-                nodesToBeKilled[childNode.nodeId] = undefined
-                childNode.setParent(parent, subPath)
-                res[index] = childNode // reuse node
-            } else {
-                // Lives somewhere else (note that instantiate might still reconcile for complex types!)
-                res[index] = childType.instantiate(parent, subPath, undefined, newValue)
-            }
-        } else if (isMutable(newValue)) {
-            // The snapshot of a tree node, try to reconcile based on id
-            const reconcilationCandidate = findReconcilationCandidates(newValue)
-            if (reconcilationCandidate) {
-                const childNode = childType.reconcile(reconcilationCandidate, newValue)
-                nodesToBeKilled[reconcilationCandidate.nodeId] = undefined
-                childNode.setParent(parent, subPath)
-                res[index] = childNode
-            } else {
-                res[index] = childType.instantiate(parent, subPath, undefined, newValue)
+        hasNewNode = i <= newValues.length - 1
+
+        // both are empty, end
+        if (!oldNode && !hasNewNode) {
+            break
+            // new one does not exists, old one dies
+        } else if (!hasNewNode) {
+            oldNode.die()
+            oldNodes.splice(i, 1)
+            i--
+            // there is no old node, create it
+        } else if (!oldNode) {
+            // check if already belongs to the same parent. if so, avoid pushing item in. only swapping can occur.
+            if (isStateTreeNode(newValue) && getStateTreeNode(newValue).parent === parent) {
+                // this node is owned by this parent, but not in the reconcilable set, so it must be double
+                fail(
+                    `Cannot add an object to a state tree if it is already part of the same or another state tree. Tried to assign an object to '${parent.path}/${newPaths[
+                        i
+                    ]}', but it lives already at '${getStateTreeNode(newValue).path}'`
+                )
             }
+            oldNodes.splice(i, 0, valueAsNode(childType, parent, "" + newPaths[i], newValue))
+            // both are the same, reconcile
+        } else if (areSame(oldNode, newValue)) {
+            oldNodes[i] = valueAsNode(childType, parent, "" + newPaths[i], newValue, oldNode)
+            // nothing to do, try to reorder
         } else {
-            // create a fresh MST node
-            res[index] = childType.instantiate(parent, subPath, undefined, newValue)
+            oldMatch = undefined
+
+            // find a possible candidate to reuse
+            for (let j = i; j < oldNodes.length; j++) {
```

Comment:
Note that this makes the reconciliation basically O(N)^2. For that reason there was in the original implementation a map of old instances build first (so performance is ~ O(N)). I think that could still be done in the new setup? Create a map `identifier => { target, stillAvailable }` ?

Reply:
Uhm, I am not good at this kind of things so I may be totally wrong :)
Was'nt also the old one O(N)^2 because findReconciliationCandidate would loop through all unused array items with an identifier (for lett attr in oldNodesBy etc...), with decreasing amount of loops each step the old nodes get used up.
I think here's basically the same thing, findReconciliationCandidate is mostly equal to node loop + areSame?